### PR TITLE
consensus query for ledger reward provenance

### PR DESCRIPTION
**Pull request by JaredCorduan**
This commit adds a new consensus query which returns a lot of data surrounding the reward calculation in the ledger. There is data concerning the overall calculation, but also detailed information about each individual stake pool's rewards and performance.

I updated the ledger package to master in order to get the CBOR instances for the reward provenance.

Resolves: CAD-2387

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 1]
```diff
@@ -285,6 +293,7 @@ querySupportedVersion = \case
     GetLedgerTip                               -> (>= v1)
     GetEpochNo                                 -> (>= v1)
     GetNonMyopicMemberRewards {}               -> (>= v1)
+    GetRewardProvenance                        -> (>= v2)
```
Comment (JaredCorduan) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:296]: @mrBliss is it okay to add this new query to version 2 like I've done? Is there anywhere else that I need to document it?

**Review by mrBliss — COMMENTED** [review]
@JaredCorduan Thanks for doing the work!

I'll push some fixes to this PR.

**Comment by mrBliss** [comment]
> I'll push some fixes to this PR.

I will probably force-push, so that the history is clean (and each commit is buildable).

**Comment by mrBliss** [comment]
@nfrisby Adding you as a reviewer, as you will be doing this in the future :slightly_smiling_face:.

Unfortunately, adding new queries requires adding a new version. See the the last commit for a template on what to do. The golden test results are generated automatically by running the golden tests.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 2]
```diff
@@ -85,6 +86,8 @@ data instance Query (ShelleyBlock era) :: Type -> Type where
   GetNonMyopicMemberRewards
     :: Set (Either SL.Coin (SL.Credential 'SL.Staking (EraCrypto era)))
     -> Query (ShelleyBlock era) (NonMyopicMemberRewards (EraCrypto era))
+  GetRewardProvenance
+    :: Query (ShelleyBlock era) (SL.RewardProvenance (EraCrypto era))
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Done.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 3]
```diff
@@ -207,6 +212,8 @@ instance SameDepIndex (Query (ShelleyBlock era)) where
     = Nothing
   sameDepIndex (GetNonMyopicMemberRewards _) _
     = Nothing
+  sameDepIndex GetRewardProvenance _
+    = Nothing
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Done.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 4]
```diff
@@ -154,6 +157,8 @@ instance ShelleyBasedEra era => QueryLedger (ShelleyBlock era) where
         GetNonMyopicMemberRewards creds ->
           NonMyopicMemberRewards $
             SL.getNonMyopicMemberRewards globals st creds
+        GetRewardProvenance ->
+          snd $ SL.getRewardInfo globals st
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Done.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs (hunk 1 above)
Comment (JaredCorduan) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: That makes sense, thank you for the explanation!

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JaredCorduan) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Yep, I agree, that's nicer.

**Comment by JaredCorduan** [comment]
> > I'll push some fixes to this PR.
> 
> I will probably force-push, so that the history is clean (and each commit is buildable).

Thank you for cleaning this up for me @mrBliss !

**Review by mrBliss — APPROVED** [review]
bors merge

**Pull request merged**